### PR TITLE
Pam 10561 use caching service

### PR DESCRIPTION
**Pull request by tobal**

### CODE REVIEW

File: default.cfg [hunk 1]
```diff
@@ -1,16 +1,4 @@
 [starling]
-# Optional: If the SPS node is joined to One Identity Starling, you do not have to
```
Comment (krajorama) [inline default.cfg:2]: I'd mention here that "This plugin is only usable if the SPS node is joined to Starling."

File: lib/plugin.py [hunk 2]
```diff
@@ -37,26 +50,21 @@ def _authentication_steps(self):
 
     def do_authenticate(self):
         client = self.construct_mfa_client()
-        return client.execute_authenticate(self.username, self.mfa_identity, self.mfa_password)
+        verdict = client.execute_authenticate(self.username, self.mfa_identity, self.mfa_password)
+        if self._user_doesnt_exist(verdict):
+            self.__cache.set(key=USER_ID_CACHE_KEY, value={'user_id': None, 'is_valid': False}, ttl=3600)
```
Comment (krajorama) [inline lib/plugin.py]: key must specify the user (email + phone number)

File: default.cfg [hunk 3]
```diff
@@ -1,16 +1,4 @@
...
-# account. Navigate to Dashboard and click Subscription Key.
-#   To use a local Credential Store to host this data, read the "Store sensitive
-# plugin data securely" section in the documentation.
-; api_key=<$-or-API-Key>
-
-# Optional: the URL where the One Identity Starling server can be accessed.
-; api_url=https://api.2fa.cloud.oneidentity.com
-
 # Optional: the timeout for push notification (seconds)
 ; timeout=60
```
Comment (krajorama) [inline default.cfg]: Need
```
[memory-cache]
prefix=oneidentity:starling
# Time to live of wrong userid?
ttl=3600
```

**Review by krajorama — CHANGES_REQUESTED** [review]

File: lib/plugin.py (hunk 2 above)
Comment (tobal) [reply lib/plugin.py]: Done

File: default.cfg (hunk 3 above)
Comment (tobal) [reply default.cfg]: Done

File: default.cfg (hunk 1 above)
Comment (tobal) [reply default.cfg:2]: Done

File: lib/plugin.py [hunk 4]
```diff
@@ -70,18 +95,41 @@ def create_push_details(self):
...
+        cached_user_info = self.__cache.get(USER_ID_CACHE_KEY_PREFIX)
+        if cached_user_info:
+            if cached_user_info['is_valid']:
+                self.mfa_identity = cached_user_info['user_id']
+                return
+            else:
+                reason = 'Cached user ID is invalid, try again later; user_id={}'.format(cached_user_info['user_id'])
+                self.logger.warning(reason)
+                return AAResponse.deny(reason=reason)
+        print(self._ldap_user_info)
```
Comment (krajorama) [inline lib/plugin.py]: debug printout should use self.logger

File: lib/plugin.py [hunk 5]
```diff
@@ -70,18 +95,41 @@ def create_push_details(self):
         return {k: v for k, v in push_details if v is not None}
 
     def _provision_user(self):
-        phone_attribute = self.plugin_configuration.get('starling_auto_provision', 'phone_attribute')
-        email_attribute = self.plugin_configuration.get('starling_auto_provision', 'email_attribute')
-        name_attribute = 'displayName'
-        if phone_attribute and email_attribute:
+        cached_user_info = self.__cache.get(USER_ID_CACHE_KEY_PREFIX)
```
Comment (krajorama) [inline lib/plugin.py]: get must use the constructed _user_id_cache_key
the whole logic here needs to be inside the "if" form line 108
if self._ldap_user_info.phone and self._ldap_user_info.email:

File: lib/starling_join_client.py [hunk 6]
```diff
@@ -19,76 +19,38 @@
...
-        tokens = self._request_token()
-        try:
-            cache_file.seek(0)
-            json.dump({
-                'expires': request_time + tokens['expires_in'],
-                'data': tokens,
-            }, cache_file)
+            return cached_access_token
+        else:
             logger.debug('Written cache of Starling access token')
```
Comment (krajorama) [inline lib/starling_join_client.py]: let's move this line before cache.set and say: Writing cache of Starling access token.
It's a bit out of context here.

File: lib/plugin.py [hunk 7]
```diff
@@ -70,18 +95,41 @@ def create_push_details(self):
...
+        required_attributes = [phone_attribute, email_attribute, name_attribute]
         ldap_service = LDAPServer.from_config(self.plugin_configuration)
-        return ldap_service.get_user_string_attributes(self.username, required_attributes)
+        ldap_info = ldap_service.get_user_string_attributes(self.username, required_attributes)
+        return LDAPInfo(ldap_info[phone_attribute], ldap_info[email_attribute], ldap_info[name_attribute])
+
+    def _user_doesnt_exist(self, verdict):
+        if verdict.get('verdict') == 'DENY':
+            try:
+                metadata = json.loads(verdict['additional_metadata'])
```
Comment (krajorama) [inline lib/plugin.py]: let's put a boolean property on the client that we can read here, let's not reload the boolean and handle REST API specific stuff like error code.

**Review by krajorama — CHANGES_REQUESTED** [review]

File: lib/plugin.py [hunk 8]
```diff
@@ -63,18 +95,41 @@ def create_push_details(self):
...
+        cached_user_info = self.__cache.get(USER_ID_CACHE_KEY_PREFIX)
+        if cached_user_info:
+            if cached_user_info['is_valid']:
+                self.mfa_identity = cached_user_info['user_id']
+                return
+            else:
+                reason = 'Cached user ID is invalid, try again later; user_id={}'.format(cached_user_info['user_id'])
+                self.logger.warning(reason)
+                return AAResponse.deny(reason=reason)
+        print(self._ldap_user_info)
```
Comment (gorosz) [inline lib/plugin.py]: Remove print or debug log

**Review by gorosz — CHANGES_REQUESTED** [review]
Please see comments

File: lib/plugin.py (hunk 4 above)
Comment (tobal) [reply lib/plugin.py]: Deleted

File: lib/plugin.py (hunk 5 above)
Comment (tobal) [reply lib/plugin.py]: Done

File: lib/starling_join_client.py (hunk 6 above)
Comment (tobal) [reply lib/starling_join_client.py]: Done

File: lib/plugin.py (hunk 8 above)
Comment (tobal) [reply lib/plugin.py]: Done

File: lib/plugin.py (hunk 7 above)
Comment (tobal) [reply lib/plugin.py]: Done

**Review by gorosz — APPROVED** [review]

File: lib/plugin.py [hunk 9]
```diff
@@ -70,18 +95,28 @@ def create_push_details(self):
...
+                                                  self._ldap_user_info.name)
+            self.__cache.set(key=self._user_id_cache_key, value={'user_id': user_id, 'is_valid': True}, ttl=0)
             self.mfa_identity = user_id or self.mfa_identity
 
-    def _query_user_ldap_information(self, required_attributes):
+    def _query_user_ldap_information(self):
+        phone_attribute = self.plugin_configuration.get('starling_auto_provision', 'phone_attribute')
+        email_attribute = self.plugin_configuration.get('starling_auto_provision', 'email_attribute')
+        name_attribute = 'displayName'
+        required_attributes = [phone_attribute, email_attribute, name_attribute]
```
Comment (krajorama) [inline lib/plugin.py]: These are not required, if phone or email attribute is not defined, then we can just fill LDAPInfo with Nones. See default config: "# To turn on auto-provisioning, you have to specify both attributes"

File: lib/tests/conftest.py [hunk 10]
```diff
@@ -55,30 +57,21 @@ def push_details():
...
-        yield StarlingClient(
-            environment=site_parameters['environment'],
-            poll_interval=0.1,
-            push_details=push_details
-        )
-        shutil.rmtree(tempdir, True)
-    else:
-        yield None
+def client(monkeypatch, gateway_fqdn, site_parameters, push_details):
+    tempdir = tempfile.mkdtemp()
```
Comment (krajorama) [inline lib/tests/conftest.py]: we can get rid of tempdir I think , IIRC it was used for testing the file based cache

**Review by krajorama — CHANGES_REQUESTED** [review]
I have two small comments, also it's time to write up the changelog I think.

File: lib/plugin.py (hunk 9 above)
Comment (tobal) [reply lib/plugin.py]: Done

File: lib/tests/conftest.py (hunk 10 above)
Comment (tobal) [reply lib/tests/conftest.py]: Done

**Review by krajorama — APPROVED** [review]
lgtm

**Pull request merged**